client/tso_dispatcher.go

```diff
@@ -271,55 +303,87 @@ tsoBatchLoop:
 				stream = nil
 				continue
 			default:
-				break streamChoosingLoop
 			}
+
+			// Check if any error has occurred on this stream when receiving asynchronously.
+			if err = stream.GetRecvError(); err != nil {
+				exit := !td.handleProcessRequestError(ctx, bo, streamURL, cancel, err)
+				stream = nil
+				if exit {
+					td.cancelCollectedRequests(batchController, invalidStreamID, errors.WithStack(ctx.Err()))
+					return
+				}
+				continue
+			}
+
+			break streamChoosingLoop
 		}
 		done := make(chan struct{})
 		dl := newTSDeadline(option.timeout, done, cancel)
 		select {
 		case <-ctx.Done():
 			// Finish the collected requests if the context is canceled.
-			batchController.finishCollectedRequests(0, 0, 0, errors.WithStack(ctx.Err()))
+			td.cancelCollectedRequests(batchController, invalidStreamID, errors.WithStack(ctx.Err()))
 			return
 		case td.tsDeadlineCh <- dl:
 		}
 		// processRequests guarantees that the collected requests could be finished properly.
-		err = td.processRequests(stream, dc, td.batchController)
-		close(done)
+		err = td.processRequests(stream, dc, batchController, done)
 		// If error happens during tso stream handling, reset stream and run the next trial.
-		if err != nil {
+		if err == nil {
+			// If the request is started successfully, the `batchController` will be put back to the pool when the
+			// request is finished (either successful or not). In this case, set the `batchController` to nil so that
+			// another one will be fetched from the pool.
+			// Otherwise, the `batchController` won't be processed in other goroutines concurrently, and it can be
+			// reused in the next loop safely.
+			batchController = nil
```

Comment:
The comment above said we have to set this `batchController` to nil either successful or not, do we need to set it to nil even if the `err` is not nil?

Reply:
I updated the comment. Hope it's more clear now.